### ISSUE TITLE
Loss Calucation

### ISSUE THREAD

**Issue by YuchenHui22314** [sanderwood/bgpt#5]
Hi.

While reviewing the code at line 200 in train-gen.py, https://github.com/sanderwood/bgpt/blob/main/train-gen.py#L200

`def process_one_batch(batch):
    input_patches, input_masks = batch
    loss = model(input_patches, input_masks).loss

    # Reduce the loss on GPU 0
    if world_size > 1:
        loss = loss.unsqueeze(0)
        dist.reduce(loss, dst=0)
        loss = loss / world_size
        dist.broadcast(loss, src=0)

    return loss

Here, why should we do loss = loss/world_size? I am wondering because according to my knowledge, the gradients will automatically be unscaled(divided) by world_size while doing loss.backword() w.r.t. a model using DDP. So if we do one another time here at line 200, the gradient would be world_size times smaller than the value it should have.

Do please correct me if I am wrong!

Thanks a lot!

Yuchen

**Comment by sanderwood** [comment]
Hi,

Sorry for the late response. The `dist.reduce(loss, dst=0)` will perform the sum operation, so we need to divide the world_size to manually unscale loss.
